ai21/clients/bedrock/_stream_decoder.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import annotations
+
+import json
+from functools import lru_cache
+from typing import Iterator, AsyncIterator
+
+import httpx
+from botocore.eventstream import EventStreamMessage
+from botocore.model import Shape
+
+from ai21.errors import StreamingDecodeError
+
+
+@lru_cache(maxsize=None)
+def get_response_stream_shape() -> Shape:
+    from botocore.model import ServiceModel
+    from botocore.loaders import Loader
+
+    loader = Loader()
+    bedrock_service_dict = loader.load_service_model("bedrock-runtime", "service-2")
+    bedrock_service_model = ServiceModel(bedrock_service_dict)
+    return bedrock_service_model.shape_for("ResponseStream")
+
+
+class AWSEventStreamDecoder:
+    def __init__(self) -> None:
+        from botocore.parsers import EventStreamJSONParser
+
+        self.parser = EventStreamJSONParser()
+
+    def iter(self, response: httpx.Response) -> Iterator[str]:
+        """Given an iterator that yields lines, iterate over it & yield every event encountered"""
+        from botocore.eventstream import EventStreamBuffer
+
+        event_stream_buffer = EventStreamBuffer()
+        last_message = None
+        for chunk in response.iter_bytes():
+            try:
+                event_stream_buffer.add_data(chunk)
+                for event in event_stream_buffer:
+                    message = self._parse_message_from_event(event)
+                    if message:
+                        # START TODO: remove the following conditions once response from Bedrock changes
+                        if last_message is not None:
+                            message = self._create_bedrock_last_chunk(
+                                chunk=last_message, bedrock_metrics_message=message
+                            )
+                        if '"finish_reason":null' not in message and last_message is None:
+                            last_message = message
```

Comment:
Let's rename to `previous_message` as `last_message`, makes it sound like its the last message period. wdyt?